File: src/main/java/dev/forcetower/breaker/JOrchestra.java
package dev.forcetower.breaker;

import dev.forcetower.breaker.model.*;
import dev.forcetower.breaker.result.Outcome;

import java.util.List;
import java.util.concurrent.Future;
import java.util.concurrent.CompletableFuture;

public class JOrchestra {
    private final Orchestra instance;

    public JOrchestra(Orchestra instance) {
        this.instance = instance;
    }

    public Future<Outcome<Person>> login() {
        var completable = new CompletableFuture<Outcome<Person>>();
        instance.login(Coroutines.getContinuation((outcome, throwable) -> {
            if (throwable != null) {
                completable.completeExceptionally(throwable);
            } else {
                completable.complete(outcome);
            }
        }));
        return completable;
    }

    public Future<Outcome<List<? extends Semester>>> semesters(long profileId) {
        var completable = new CompletableFuture<Outcome<List<? extends Semester>>>();
        instance.semesters(profileId, Coroutines.getContinuation((outcome, throwable) -> {
            if (throwable != null) {
                completable.completeExceptionally(throwable);
            } else {
                completable.complete(outcome);
            }
        }));
        return completable;
    }

    public Future<Outcome<List<? extends SemesterInformation>>> allSemestersInformation(long profileId) {
        var completable = new CompletableFuture<Outcome<List<? extends SemesterInformation>>>();
        instance.allSemestersInformation(profileId, Coroutines.getContinuation((outcome, throwable) -> {
            if (throwable != null) {
                completable.completeExceptionally(throwable);
            } else {
                completable.complete(outcome);
            }
        }));
        return completable;
    }

    public Future<Outcome<List<? extends DisciplineData>>> grades(long profileId, long semesterId) {
        var completable = new CompletableFuture<Outcome<List<? extends DisciplineData>>>();
        instance.grades(profileId, semesterId, Coroutines.getContinuation((outcome, throwable) -> {
            if (throwable != null) {
                completable.completeExceptionally(throwable);
            } else {
                completable.complete(outcome);
            }
        }));
        return completable;
    }

    public Future<Outcome<MessagesDataPage>> messages(long profileId) {
        return messages(profileId, "");
    }

    public Future<Outcome<MessagesDataPage>> messages(long profileId, String until) {
        var completable = new CompletableFuture<Outcome<MessagesDataPage>>();
        instance.messages(profileId, until, Coroutines.getContinuation((outcome, throwable) -> {
            if (throwable != null) {
                completable.completeExceptionally(throwable);
            } else {
                completable.complete(outcome);
            }
        }));
        return completable;
    }

    public Future<Outcome<List<? extends Lecture>>> lectures(long classId, int limit, int offset) {
        var completable = new CompletableFuture<Outcome<List<? extends Lecture>>>();
        instance.lectures(classId, limit, offset, Coroutines.getContinuation((outcome, throwable) -> {
            if (throwable != null) {
                completable.completeExceptionally(throwable);
            } else {
                completable.complete(outcome);
            }
        }));
        return completable;
    }

    public Future<Outcome<List<? extends LectureMissed>>> absences(long profileId,long classId, int limit, int offset) {
        var completable = new CompletableFuture<Outcome<List<? extends LectureMissed>>>();
        instance.absences(profileId, classId, limit, offset, Coroutines.getContinuation((outcome, throwable) -> {
            if (throwable != null) {
                completable.completeExceptionally(throwable);
            } else {
                completable.complete(outcome);
            }
        }));
        return completable;
    }
}

File: src/main/kotlin/dev/forcetower/breaker/JOrchestra.kt
package dev.forcetower.breaker

import dev.forcetower.breaker.model.*
import dev.forcetower.breaker.result.Outcome
import kotlinx.coroutines.DelicateCoroutinesApi
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.future.future
import java.util.concurrent.CompletableFuture

@OptIn(DelicateCoroutinesApi::class)
class JOrchestra constructor(private val instance: Orchestra) {
    fun setAuthorization(authorization: Authorization) {
        instance.setAuthorization(authorization)
    }

    fun login(): CompletableFuture<Outcome<Person>> {
        return GlobalScope.future { instance.login() }
    }

    fun semesters(profileId: Long): CompletableFuture<Outcome<List<Semester>>> {
        return GlobalScope.future { instance.semesters(profileId) }
    }

    fun allSemestersInformation(profileId: Long): CompletableFuture<Outcome<List<SemesterInformation>>> {
        return GlobalScope.future { instance.allSemestersInformation(profileId) }
    }

    fun grades(profileId: Long, semesterId: Long): CompletableFuture<Outcome<List<DisciplineData>>> {
        return GlobalScope.future { instance.grades(profileId, semesterId) }
    }

    fun messages(profileId: Long, until: String = ""): CompletableFuture<Outcome<MessagesDataPage>> {
        return GlobalScope.future { instance.messages(profileId, until) }
    }

    fun lectures(classId: Long, limit: Int, offset: Int): CompletableFuture<Outcome<List<Lecture>>> {
        return GlobalScope.future { instance.lectures(classId, limit, offset) }
    }

    fun absences(profileId: Long, classId: Long, limit: Int, offset: Int): CompletableFuture<Outcome<List<LectureMissed>>> {
        return GlobalScope.future { instance.absences(profileId, classId, limit, offset) }
    }
}